File: src/custos/assistants/rule_policy.py
"""A7 ``rule-policy`` assistant - pure deterministic rules, no LLM .

Fast path for low-risk read ops. Takes a rule table at construction (a list
of ``(match, AssistantOutput)`` pairs); ``decide`` evaluates first-match-wins
using :class:`~custos.policy.match.MatchSpec` and returns the matching rule's
output. If no rule matches, returns a ``deny`` with risk scaled from the
tool's risk_tier.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence

from custos.assistants.base import AssistantBase
from custos.policy.match import MatchSpec
from custos.schema import AssistantOutput, Decision, Invocation, SubjectContext

__all__ = ["RulePolicy"]


# Static risk score for an unmatched call: tool_tier / 5.0, clamped to [0,1).
_TIER_RISK_FACTOR = 0.2


class RulePolicy(AssistantBase):
    """Pure deterministic rules; no LLM (A7)."""

    name = "rule-policy"

    def __init__(
        self,
        rules: Sequence[tuple[Mapping[str, object], AssistantOutput]] | None = None,
        *,
        default_decision: Decision = Decision.DENY,
    ) -> None:
        self._compiled: list[tuple[MatchSpec, AssistantOutput]] = [
            (MatchSpec.from_mapping(m), out) for m, out in (rules or [])
        ]
        self._default_decision = default_decision

    def decide(self, inv: Invocation, ctx: SubjectContext) -> AssistantOutput:
        for match, out in self._compiled:
            if match.matches(inv):
                return out
        # No rule matched: deny with risk scaled from the tool's tier.
        tier = inv.descriptor.risk_tier if inv.descriptor else 3
        risk = min(1.0, max(0.0, tier * _TIER_RISK_FACTOR))
        return AssistantOutput(
            decision=self._default_decision,
            risk=risk,
            reasoning=f"rule-policy: no rule matched; tier={tier}",
        )

File: packages/custos-ts/src/assistants/rule_policy.ts
// A7 — `rule-policy`. Pure deterministic rules; no LLM. Fast path for
// low-risk read ops. Re-evaluates the invocation against an inner
// `Policy` (the "rule table") and returns the inner verdict.

// Per D17 the TS `@taqiy/custos-core` ships A7 in-process; LLM-backed
// assistants route via the sidecar.

import type { Assistant } from "./base.ts";
import type { AssistantOutput, Invocation, SubjectContext } from "../schema.ts";
import { Policy } from "../policy/engine.ts";

export class RulePolicyAssistant implements Assistant {
  readonly name = "rule-policy";
  readonly exfiltratesArgs = false;
  private readonly ruleTable: Policy;

  constructor(ruleTable: Policy) {
    this.ruleTable = ruleTable;
  }

  decide(inv: Invocation, _ctx: SubjectContext): AssistantOutput {
    const evalResult = this.ruleTable.evaluate(inv);
    switch (evalResult.outcome) {
      case "allow":
        return {
          decision: "allow_once",
          risk: 0,
          reasoning: "A7 rule-policy: inner rule-table allow",
          fatigue_hint: false,
          persist_rule: null,
        };
      case "deny":
        return {
          decision: "deny",
          risk: 1.0,
          reasoning: "A7 rule-policy: inner rule-table deny",
          fatigue_hint: false,
          persist_rule: null,
        };
      case "prompt":
        return {
          decision: "prompt",
          risk: 0.5,
          reasoning: "A7 rule-policy: inner rule-table prompt",
          fatigue_hint: false,
          persist_rule: null,
        };
      case "assist":
        // Inner rule-table `assist:<name>` would re-enter the assistant
        // pipeline; the gateway handles assistant chaining separately. A7
        // treats an inner ASSIST as a deny-then-prompt by default.
        return {
          decision: "prompt",
          risk: 0.5,
          reasoning: "A7 rule-policy: inner ASSIST reified as prompt",
          fatigue_hint: false,
          persist_rule: null,
        };
    }
  }
}